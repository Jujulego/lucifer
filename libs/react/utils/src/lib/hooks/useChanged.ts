import isEqual from 'lodash.isequal';

import usePrevious from './usePrevious';

// Hook
function useChanged<T>(obj: T): boolean {
  return isEqual(obj, usePrevious<T>(obj));
}

export default useChanged;
