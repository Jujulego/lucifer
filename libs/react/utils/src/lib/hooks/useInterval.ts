import { useEffect, useState } from 'react';
import dayjs from 'dayjs';
import duration, { DurationUnitType } from 'dayjs/plugin/duration';

dayjs.extend(duration);

// Hooks
function useInterval(every: number, unit?: DurationUnitType): number {
  // State
  const [count, setCount] = useState(0);

  // Effects
  useEffect(() => {
    const ms = dayjs.duration(every, unit).milliseconds();
    const interval = setInterval(() => setCount(old => old + 1), ms);

    return () => { clearInterval(interval); }
  }, [every, unit]);

  return count;
}

export default useInterval;
