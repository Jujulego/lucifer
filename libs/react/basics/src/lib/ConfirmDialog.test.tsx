import React, { useEffect } from 'react';
import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { createMount, createShallow } from '@material-ui/core/test-utils';

import ConfirmDialog from './ConfirmDialog';
import { useConfirm } from './confirm.hooks';

// Setup
let mount: ReturnType<typeof createMount>;
let shallow: ReturnType<typeof createShallow>;

beforeAll(() => {
  mount = createMount();
  shallow = createShallow();
});

afterAll(() => {
  mount.cleanUp();
});

// Tests
describe('ConfirmDialog', () => {
  it('should be closed and empty', () => {
    const TestComponent = () => {
      const { state } = useConfirm("");

      return (
        <ConfirmDialog state={state}>
          {(txt) => txt}
        </ConfirmDialog>
      )
    };

    // Render
    const wrapper = shallow(<TestComponent />).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('should be opened with content', () => {
    const TestComponent = () => {
      const { state, confirm } = useConfirm("");
      if (!state.open) confirm("Test");

      return (
        <ConfirmDialog state={state}>
          {(txt) => txt}
        </ConfirmDialog>
      )
    };

    // Render
    const wrapper = shallow(<TestComponent />).dive();

    expect(wrapper).toMatchSnapshot();
  });
});

describe('useConfirm().confirm', () => {
  // Setup
  let spy: jest.Mock<void,[boolean]>;

  let release: () => void;
  let waiter: Promise<void>;

  let wrapper: ReactWrapper;

  beforeEach(() => {
    // Utils
    spy = jest.fn();
    waiter = new Promise<void>(resolve => {
      release = resolve;
    });

    // Component
    const TestComponent = () => {
      const { state, confirm } = useConfirm("");

      useEffect(() => {
        confirm("Test")
          .then(spy)
          .finally(release);
      }, [confirm]);

      return (
        <ConfirmDialog state={state}>
          { (txt) => txt }
        </ConfirmDialog>
      )
    };

    // Render
    wrapper = mount(<TestComponent />);
  });

  // Tests
  it('should return false if canceled', async () => {
    // should have 2 buttons
    const buttons = wrapper.find('button');
    expect(buttons).toHaveLength(2);

    // 1st should contain "Annuler"
    const btn = buttons.first();
    expect(btn.text()).toBe('Annuler');

    // Act event
    act(() => {
      btn.simulate('click');
    });

    await waiter;
    expect(spy).toBeCalledWith(false);
  });

  it('should return true if confirmed', async () => {
    // should have 2 buttons
    const buttons = wrapper.find('button');
    expect(buttons).toHaveLength(2);

    // 2nd should contain "Confirmer"
    const btn = buttons.last();
    expect(btn.text()).toBe('Confirmer');

    // Act event
    act(() => {
      btn.simulate('click');
    });

    await waiter;
    expect(spy).toBeCalledWith(true);
  });
});
