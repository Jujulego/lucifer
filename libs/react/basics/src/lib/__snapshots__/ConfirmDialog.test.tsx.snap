// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog should be closed and empty 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={false}
>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      color="secondary"
    >
      Annuler
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      color="primary"
    >
      Confirmer
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`ConfirmDialog should be opened with content 1`] = `
<WithStyles(ForwardRef(Dialog))
  onClose={[Function]}
  open={true}
>
  Test
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      color="secondary"
      onClick={[Function]}
    >
      Annuler
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      onClick={[Function]}
    >
      Confirmer
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
